src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/VirtualDocumentFactoryBase.cs

```diff
@@ -0,0 +1,139 @@
+            }
+
+            if (!hostDocumentBuffer.ContentType.IsOfType(HostDocumentContentTypeName))
+            {
+                // Another content type we don't care about.
+                virtualDocument = null;
+                return false;
+            }
+
+            var hostDocumentUri = _fileUriProvider.GetOrCreate(hostDocumentBuffer);
+
+            // E.g. Index.cshtml => Index.cshtml__virtual.html (for html), similar for other languages
+            var virtualLanguageFilePath = hostDocumentUri.GetAbsoluteOrUNCPath() + LanguageFileNameSuffix;
+            var virtualLanguageUri = new Uri(virtualLanguageFilePath);
+
+            var languageBuffer = _textBufferFactory.CreateTextBuffer();
+            _fileUriProvider.AddOrUpdate(languageBuffer, virtualLanguageUri);
+            languageBuffer.Properties.AddProperty(LSPConstants.ContainedLanguageMarker, true);
+
+            if (!(AdditionalLanguageBufferProperties is null))
+            {
+                foreach (KeyValuePair<object, object> keyValuePair in AdditionalLanguageBufferProperties)
+                {
+                    languageBuffer.Properties.AddProperty(keyValuePair.Key, keyValuePair.Value);
+                }
+            }
+
+            // Create a text document to trigger language server initialization for the contained language.
+            _textDocumentFactory.CreateTextDocument(languageBuffer, virtualLanguageFilePath);
+
+            languageBuffer.ChangeContentType(LanguageLSPContentType, editTag: null);
+
+            virtualDocument = CreateVirtualDocument(virtualLanguageUri, languageBuffer);
+            return true;
+        }
+
+        /// <summary>
+        /// Creates and returns specific virtual document instance
+        /// </summary>
+        /// <param name="uri">Virtual document URI</param>
+        /// <param name="textBuffer">Language text buffer</param>
+        /// <returns>Language-specific virtual document instance</returns>
+        protected abstract VirtualDocument CreateVirtualDocument(Uri uri, ITextBuffer textBuffer);
+
+        /// <summary>
+        /// Returns contained language content type name that's registered with IContentTypeRegisteryService
+        /// </summary>
+        protected abstract string LanguageContentTypeName { get; }
+
+        /// <summary>
+        /// Returns contained language uri suffix, e.g. __virtual.html or __virtual.css
+        /// </summary>
+        protected abstract string LanguageFileNameSuffix { get; }
+
+        /// <summary>
+        /// Returns additional properties (if any) to set on the language text buffer prior to language server init
+        /// </summary>
+        protected virtual Dictionary<object, object> AdditionalLanguageBufferProperties => null;
+
+        /// <summary>
+        /// Returns supported host document content type name (i.e. host document content type of the for which this factory can create virtual documents)
```

Comment:
```suggestion
        /// Returns supported host document content type name (i.e. host document content type for which this factory can create virtual documents)
```

Reply:
Yep, makes sense.

---
In reply to: [451978930](https://github.com/dotnet/aspnetcore-tooling/pull/2201#discussion_r451978930) [](ancestors = 451978930)